src/Blocks/BlockEndPortalFrame.h

```diff
@@ -0,0 +1,239 @@
+
+#pragma once
+
+#include "BlockHandler.h"
+
+
+
+
+
+class cBlockEndPortalFrameHandler :
+	public cMetaRotator<cBlockHandler, 0x03, 0x00, 0x01, 0x02, 0x03, true>
+{
+public:
+	cBlockEndPortalFrameHandler(BLOCKTYPE a_BlockType):
+		cMetaRotator<cBlockHandler, 0x03, 0x00, 0x01, 0x02, 0x03, true>(a_BlockType)
+	{
+	}
+
+
+
+
+
+	virtual bool GetPlacementBlockTypeMeta(
+		cChunkInterface & a_ChunkInterface, cPlayer & a_Player,
+		int a_BlockX, int a_BlockY, int a_BlockZ, eBlockFace a_BlockFace,
+		int a_CursorX, int a_CursorY, int a_CursorZ,
+		BLOCKTYPE & a_BlockType, NIBBLETYPE & a_BlockMeta
+	) override
+	{
+		a_BlockType = m_BlockType;
+		a_BlockMeta = cBlockRedstoneRepeaterHandler::RepeaterRotationToMetaData(a_Player.GetYaw());
+		return true;
+	}
+
+
+
+
+
+	virtual void OnPlaced(cChunkInterface & a_ChunkInterface, cWorldInterface & a_WorldInterface, int a_BlockX, int a_BlockY, int a_BlockZ, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta) override
+	{
+		/*
+		PORTAL FINDING ALGORITH
+		=======================
+		- Get clicked base block
+		- Check diagonally (clockwise) for another portal block
+			- if exists, and has eye, Continue.
+				Abort if any are facing the wrong direction.
+			- if doesn't exist, check horizontally (the block to the left of this block).
+				Abort if there is no horizontal block.
+		- After a corner has been met, circle around each wall, ensuring valid portal frames connect the rectangle.
+		- Track the NorthWest Corner, and the dimensions.
+		- If dimensions are valid, create the portal.
+		*/
+		// 0x4 is the bit which signifies the eye of ender is in it.
+		// LOG("PortalPlaced, meta %d", a_BlockMeta);
+		if ((a_BlockMeta & 4) == 4)
+		{
+			// LOG("Location is %d %d %d", a_BlockX, a_BlockY, a_BlockZ);
+			// Direction is the first two bits, masked by 0x3
+			FindAndSetPortal(Vector3i(a_BlockX, a_BlockY, a_BlockZ), a_BlockMeta & 3, a_ChunkInterface, a_WorldInterface);
+		}
+	}
+
+
+
+
+
+	enum EndPortalReturns
+	{
+		E_VALID_PORTAL_FRAME = 0,
+		E_INVALID_PORTAL_FRAME = 1,
+		E_NOT_PORTAL_FRAME = 2,
+	};
+
+
+
+
+	/** Returns 0 if portal cannot be made, 1 if portal was made. */
+	bool FindAndSetPortal(Vector3i a_FirstFrame, NIBBLETYPE a_Direction, cChunkInterface & a_ChunkInterface, cWorldInterface & a_WorldInterface)
+	{
+		const int MIN_PORTAL_WIDTH = 3;
+		const int MAX_PORTAL_WIDTH = 4;
+
+		
+		Vector3i Left[4] =
```

Comment:
Constructing these every time seems silly to me, since these coord adjustments don't change with the portal in question. Should these be globalised, as in the [fire simulator](https://github.com/cuberite/cuberite/blob/6309c6a97fdbabfde978358f5f9a0f61ab74f91f/src/Simulator/FireSimulator.cpp#L42)

Reply:
Globally, they wouldn't be too useful, since the orientation of the portal blocks is not the same as the orientation of all other blocks. But I can make them static const.